Modules/_ctypes/_ctypes.c

```diff
@@ -3774,19 +3792,37 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
         return NULL;
     }
 #else
+	/* dlerror() always returns the latest error.
+	 *
+	 * Clear the previous value before calling dlsym(),
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+    (void)dlerror();
     address = (PPROC)dlsym(handle, name);
-    if (!address) {
 #ifdef __CYGWIN__
-/* dlerror() isn't very helpful on cygwin */
+    if (!address) {
+        /* dlerror() isn't very helpful on cygwin */
         PyErr_Format(PyExc_AttributeError,
                      "function '%s' not found",
                      name);
+        Py_DECREF(ftuple);
+        return NULL;
+    }
 #else
-        PyErr_SetString(PyExc_AttributeError, dlerror());
-#endif
+    const char *dlerr = dlerror();
+    if (dlerr) {
```

Comment:
For the case where another thread sets `dlerror` at the wrong moment, I'd prefer keeping `!address` as the condition here. We should only look at `dlerr` if we actually got NULL.

Reply:
Ok, I see your point.

Even though glibc's `dlerror()` is thread safe, it doesn't really matter to us.

So, let's flip the `if`s, and first check for NULL address, then for `dlerr`.

In this way, and given that our `#ifdefs` are not granular enough (esp. in this case where we are in an `#else`), we won't get bitten on a system where `dlerror()` is not thread-safe.

Of course, we could always print a wrong message, but that is just as possible even if we don't flip the `if`s.

WDYT?